Marsiple/App/New Group/PhotoCollectionViewController.swift

```diff
@@ -47,4 +59,32 @@ private extension PhotoCollectionViewController {
             flowLayout.sectionInset = UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 10)
         }
     }
+    
+    func setupActivityIndicatorView() {
+        activityIndicatorView.startAnimating()
+        activityIndicatorView.activityIndicatorViewStyle = .gray
+        view.addSubview(activityIndicatorView)
+        activityIndicatorView.snp.makeConstraints {
+            $0.edges.equalTo(view.safeAreaLayoutGuide)
+        }
+    }
+    
+    func downloadThumbnails() {
+        let group = DispatchGroup()
+        photos.forEach { photo in
+            group.enter()
+            DataFetcher.getThumbnail(withUrl: photo.thumbnailUrl,
+                                          success: { [weak self] image in
+                                            self?.thumbnails.append(image)
```

Comment:
I don't like the fact that you maintain two data sources here - the `photos` and the `thumbnails`. I would suggest not downloading the images first, but, based on a `photo` at a certain `indexPath`, send the `thumbnailURL` into the `cell` and let the `cell` fire off the request. That way you don't keep a separate image cache in memory. 

You could also add an `activity indicator` for each cell, showing whether the `cell`'s currently downloading the image.

You can add the [`Kingfisher`](https://github.com/onevcat/Kingfisher) CocoaPod for image downloads and caching, as it's one of the (if not the) libraries for this purpose.

Reply:
I taught cells shouldn't do any business logic, so I just gave them data to present.

Reply:
You thought right - this is one of those cases where the practicality outweighs the cost. 

If using `Kingfisher`, the call-site will look quite simple and the overall logic of the app will simplify.